### DISCUSSION TITLE
Add support for external JWT public keys

### DISCUSSION THREAD

**Discussion by zs-ko** [directus/directus#18571] [Feature Requests]
### Summary

When using external authentication systems like Auth0/Okta or similar we could leverage that as our authentication and user database and only have directus read the roles and user information from the signed JWT. But then Directus would need to support custom public keys/external JWT. 

### Motivation

Keeping the tokens refresh/access in sync when running a web page with ssr is almost impossible. Tokens times out and looses sync depending on client/server requests. 

In stead of handling the authentication part in Directus we would like to use external authentication systems capable of delivering the required information in the JWT token. 

This would allow us to better control the access to directus application by leveraging a signed token from a third party authentication service.

### Detailed Design

Add an option to read trusted public keys from a URL and an array. Many external auth providers have a dedicated JWT endpoint to retrieving tokens from a jwks_uri.

By adding a flag to external providers like
AUTH_GOOGLE_TRUSTED="true"
AUTH_GOOGLE_JWKS_URI=""
AUTH_GOOGLE_JWKS_PUBLIC_KEYS="somekey,someotherkey"

### Must Have:
- A Add public keys as strusted

### Should Have:
- B Read public keys from JWKS URL

### Could Have:
- C Read public keys from JWKS URL by getting it from well-known endpoint

### Drawbacks

one would need to keep token requirements documented as they change

### Alternatives

no alternative it would have to be built into the core of application

### Adoption Strategy

Should not cause any breaking changes. 
It would be really easy for users to take advantage off and they would have to enable a flag before it would even function

**Comment by zs-ko** [comment]
this could also solve other scenarioes like mentioned in 
[17840](https://github.com/directus/directus/discussions/17840)\
[17947](https://github.com/directus/directus/discussions/17947)\
[4019](https://github.com/directus/directus/discussions/4019)\
\
Not exactly the same, but providing a unified jwk could handle such use cases.

**Comment by Zyles** [comment]
I would like to use Amazon Cognito for the user pool for global admins across multiple Directus projects, so that would be great.

**Comment by zs-ko** [comment]
i looked into this and seems like this can be resovled with a plugin. Currently have an early version of it under development, will release it a soon as i feel its good enough.

**Comment by zs-ko** [comment]
will consider this really alpha for now, not 100% completed or tested yet, but getting there.

[External JWT Extension](https://github.com/Zerosubnet/directus-extension-external-jwt)
 [NPM](https://www.npmjs.com/package/@zerosubnet/directus-extension-external-jwt)

**Comment by rijkvanzanten** [comment]
@zs-ko How would this tie in with the studio app? Would you imagine the user would log in through SSO which then generates and uses the same token you'd pass directly?

**Reply by zs-ko** [reply]
@rijkvanzanten  sorry for the late reply here. I figure that we could trust external JWT providers and define methods for extracting and validating the JWT tokens.

Right now directus is generating a token that is then verified localy. my suggestion is to implement and external authentication system so that externaly signed or generated tokens can be validated.

Currently we are using Zitadel to handle authentication and can leverage either validating the jwt directly or even better using the introspection endpoint provided from oauth2 spec we can validate the token on a request and then cache it for x amount of time so not to hammer the introspection endpoint.



If you have a look at these extensions that implements some of the features
this is the one i created and it currently is not using introspection but rather blindly reading the JWT and validating it's signature against the JSON Web Key Sets. 

When validating against an introspection endpoint the issued token do not have to be a jwt and will also perform additional checks like early revocation and so on.
https://zitadel.com/docs/apis/openidoauth/endpoints#introspection_endpoint

My repo
https://github.com/zerosubnet/directus-extension-external-jwt/tree/beta

othe repo firebase auth
https://github.com/workatease/directus-extension-firebase-user-auth/blob/master/src/index.ts

so i guess at a minimum it should support JWT tokens since introspection is not supported for like Auth0 see https://community.auth0.com/t/introspection-endpoint-for-opaque-tokens-or-more-flexible-rules-to-get-clear-jwt-access-token/63866/5

And secondary support introspection. This would allow a frontend to request a single or multiple tokens from an external Authentication provider and use it against directus api and other services.

This would allow for something like the following without making multiple authentication redirects.

* Client request token from auth provider
* Auth provider returns jwt
* Client sends jwt to directus and request a collection
* Client reads collection and sends request to webshop api with same token

**Comment by JamesW1** [comment]
Update on this subject - as of v11.14 we've exposed some utilities to make building custom JWT validation extensions easier. Rather than implementing provider-specific functionality in core, we've taken a flexible, extension-based approach.

This lets you build custom JWT validation tailored to your provider and requirements, giving you control over:

- How you validate tokens (JWKS, introspection, or direct validation)
- Which providers you integrate with (Auth0, Okta, Zitadel, Cognito, Firebase, etc.)
- How you extract and map claims to Directus users and roles
- Token caching strategies and validation logic

We wanted to provide the building blocks and guidance rather than rigid provider integrations - while JWTs follow standard specifications, the custom scopes and claims organizations use vary widely based on their domain model.

Reference tutorial here: https://directus.io/docs/tutorials/extensions/validating-third-party-jwts-in-directus
